rfcs/0134-nix-store-layer.md

```diff
@@ -0,0 +1,271 @@
+The oil shell blog as a [great post](https://www.oilshell.org/blog/2022/02/diagrams.html) with more details on the concept.
+It's not every day that a project happens upon a great narrow waist design, but I believe we've discovered a very good one with Nix, and that should be seen as a *key asset*, even if it is not how we recruit "regular users".
+
+By making a store-only Nix, we put more emphasis on this key interface.
+All functionality the store-only Nix offers factors through this interface.
+The upper half of Nix likewise uses the lower half through this interface.
+The daemon protocol represents this interface for IPC, and allows either half to be swapped for a different implementation.
+
+To help explain the community-building benefits, it might help to go over some specific examples.
+
+### Tvix and go-nix
+
+In https://tvl.fyi/blog/rewriting-nix, TVL announced that, frustrated in trying to refactor Nix into something more modular and flexible, they were aiming to make a new implementation from scratch.
+More recently, in https://tvl.fyi/blog/tvix-status-september-22 they lay out a basic approach of two projects:
+
+- [*Tvix*](https://cs.tvl.fyi/depot/-/tree/tvix) is a new implementation of the Nix language evaluator,
+
+- [*go-nix*](https://github.com/nix-community/go-nix) is a new implementation of the store layer.
+
+First of all, the fact that they are planning on two completely separate implementation oriented around this same "narrow waist" is testament to the appeal of the design.
+
+Second of all, note that per they have separate, orthogonal experiments they wish to run on both sides of the store interface divide.
+Above, they want to experiment with radically different evaluation strategies, especially to speed up Nixpkgs evaluation.
+Below, they want to experiment with the standardized containerization technologies that already exist for new ways of sandboxing and distributing builds with less bespoke Nix-specific code.
+They also want to apply the layering paradigm *within* go-nix, fostering even more modularity.
+
+I think these are great goals, and for the sake of the ecosystem as a whole, it should be as easy as possible to run such new experiments.
+In particular, a novel evaluator should be usable with the standard C++ Nix store layer, and a novel store layer should be reusable with the standard C++ Nix evaluator.
+
+Yes, strictly speaking, we only need a stable daemon protocol to accommodate that goal, which we have.
+But ensuring the C++ Nix components can be built separately for use in isolation further send the message that such experimentation is *good* and *welcomed*.
+Messaging matters, and making our layered architecture "official" as this RFC proposes I think sends a better message.
+
+### Guix
+
+Guix is more diverged from Nix than Tvix + go-nix, and thus hints more at the end breadth of the design space yet to be explored.
```

Comment:
TVL has taught me an excellent software engineering trick: if you want to ensure a clean separation between two parts of your codebase, write the two parts in two different languages whose communities see each other as competitors.

Absolutely brilliant; I tip my hat to these folks for figuring that out. :)